src/common/plist.mligo

```diff
@@ -0,0 +1,146 @@
+            (bad_state, "no proposal") : proposal_doubly_linked_list option)
+    | Some plist ->
+        if (plist.first = key) then
+          match Map.find_opt (plist.first, next) plist.map with
+            | Some next_key ->
+                Some
+                  { plist with
+                    first = next_key // Make the next as the first key
+                  ; map =
+                      Map.remove (next_key, prev) // Remove the previous link of the next key.
+                        (Map.remove (plist.first, next) plist.map) // Remove the first key link.
+                  }
+            | None ->
+                if (plist.first = plist.last) then
+                  // Only 1 proposal exist.
+                  (None : proposal_doubly_linked_list option)
+                else
+                  ([%Michelson ({| { FAILWITH } |} : (nat * string) -> (proposal_doubly_linked_list option))]
+                      (bad_state, "The next of the first key should exist.") : (proposal_doubly_linked_list option))
+        else if (plist.last = key) then
+          let prev_key =
+            match Map.find_opt (plist.last, prev) plist.map with
+              | Some p -> p
+              | None -> ([%Michelson ({| { FAILWITH } |} : (nat * string) -> proposal_key)]
+                  (bad_state, "Expect prev link of the last key to exist") : proposal_key) in
+          Some
+            { plist with
+              last = prev_key // Make the prev as the new last key
+            ; map =
+                Map.remove (prev_key, next) // Remove the next link of the prev key
+                  (Map.remove (plist.last, prev) plist.map) // Remove the prev link of the last key
+            }
+        else (
+          let prev_key =
+            match Map.find_opt (key, prev) plist.map with
+              | Some p -> p
+              | None -> ([%Michelson ({| { FAILWITH } |} : (nat * string) -> proposal_key)]
+                  (bad_state, "Expect prev link of a key to exist") : proposal_key) in
+          let next_key =
+            match Map.find_opt (key, next) plist.map with
+              | Some p -> p
+              | None -> ([%Michelson ({| { FAILWITH } |} : (nat * string) -> proposal_key)]
+                  (bad_state, "Expect next link of a key to exist") : proposal_key) in
+          Some { plist with
+            map =
+              (Map.add (next_key, prev) prev_key // Update the next link
+                (Map.add (prev_key, next) next_key // Update the previous link
+                  (Map.remove (key, next) // Remove the next link of the key.
+                    (Map.remove (key, prev) plist.map) // Remove the prev link of the key.
+                  )
+                )
+              )
+          }
+        )
+
+
+
+[@inline]
+let plist_update_head (key, plist_o: proposal_key * proposal_doubly_linked_list option): proposal_doubly_linked_list option =
```

Comment:
AFAIU this sets the given key as head, but why do we need this function?

I thought that we only needed to `pop` the head of the list, not put it back. 

Could you explain why this is needed in `flush_one` please?

Reply:
Probably the naming is off a bit, but what the function does here is like `pop until reaching a proposal key`.

This is no longer necessary though, I've added a normal pop function, and update the flush implementation according to your suggestion below.